subprojects/core/src/main/java/org/gradle/api/internal/DefaultNamedDomainObjectCollection.java

```diff
@@ -103,8 +103,18 @@ protected DefaultNamedDomainObjectCollection(Class<? extends T> type, ElementSou
     }
 
     // should be protected, but use of the class generator forces it to be public
-    public DefaultNamedDomainObjectCollection(DefaultNamedDomainObjectCollection<? super T> collection, CollectionFilter<T> filter, Instantiator instantiator, Namer<? super T> namer) {
-        this(filter.getType(), collection.filteredStore(filter), collection.filteredEvents(filter), collection.filteredIndex(filter), instantiator, namer);
+    public DefaultNamedDomainObjectCollection(DefaultNamedDomainObjectCollection<? super T> collection, CollectionFilter<T> elementFilter, Instantiator instantiator, Namer<? super T> namer) {
+        this(collection, Specs.satisfyAll(), elementFilter, instantiator, namer);
+    }
+
+    protected DefaultNamedDomainObjectCollection(
+        DefaultNamedDomainObjectCollection<? super T> collection,
+        Spec<String> nameFilter,
```

Comment:
❓ Why is `nameFilter` a separate thing in the collection?  Why wouldn't this be a part of `CollectionFilter`?  In other words, why treat this differently than the object type which we store in `CollectionFilter`?  Or is there a good reason why it isn't?  Seems like moving it in there would remove the need to change all these constructors and methods in subclasses.

Reply:
The important difference is that filtering on the names is simple to do properly for both realized and unrealized types and doesn't suffer from the awkwardness of testing against the value of unrealized types. Handling it separately can make that testing cleaner/simpler/cheaper.